File: kirun-js/src/engine/json/schema/array/ArraySchemaType.ts
import { isNullValue } from '../../../util/NullCheck';
import { Schema } from '../Schema';

export class ArraySchemaType {
    private singleSchema: Schema | undefined;
    private tupleSchema: Schema[] | undefined;

    public constructor(ast?: ArraySchemaType) {
        if (!ast) return;
        this.singleSchema = ast.singleSchema ? new Schema(ast.singleSchema) : undefined;
        this.tupleSchema = ast.tupleSchema ? ast.tupleSchema.map((e) => new Schema(e)) : undefined;
    }

    public setSingleSchema(schema: Schema): ArraySchemaType {
        this.singleSchema = schema;
        return this;
    }

    public setTupleSchema(schemas: Schema[]): ArraySchemaType {
        this.tupleSchema = schemas;
        return this;
    }

    public getSingleSchema(): Schema | undefined {
        return this.singleSchema;
    }

    public getTupleSchema(): Schema[] | undefined {
        return this.tupleSchema;
    }

    public isSingleType(): boolean {
        return !isNullValue(this.singleSchema);
    }

    public static of(...schemas: Schema[]): ArraySchemaType {
        if (schemas.length == 1) return new ArraySchemaType().setSingleSchema(schemas[0]);

        return new ArraySchemaType().setTupleSchema(schemas);
    }

    public static from(obj: any): ArraySchemaType | undefined {
        if (!obj) return undefined;
        if (Array.isArray(obj))
            return new ArraySchemaType().setTupleSchema(Schema.fromListOfSchemas(obj)!);

        let x = Schema.from(obj);
        if (!x) return undefined;
        return new ArraySchemaType().setSingleSchema(x);
    }
}

File: kirun-py/src/kirun_py/json/schema/array/array_schema_type.py
from __future__ import annotations
from typing import Optional, List, TYPE_CHECKING

if TYPE_CHECKING:
    from kirun_py.json.schema.schema import Schema


class ArraySchemaType:
    def __init__(self, ast: Optional[ArraySchemaType] = None):
        self._single_schema: Optional[Schema] = None
        self._tuple_schema: Optional[List[Schema]] = None

        if ast is not None:
            from kirun_py.json.schema.schema import Schema as SchemaClass
            if ast._single_schema is not None:
                self._single_schema = SchemaClass(ast._single_schema)
            if ast._tuple_schema is not None:
                self._tuple_schema = [SchemaClass(e) for e in ast._tuple_schema]

    def set_single_schema(self, schema: Schema) -> ArraySchemaType:
        self._single_schema = schema
        return self

    def set_tuple_schema(self, schemas: List[Schema]) -> ArraySchemaType:
        self._tuple_schema = schemas
        return self

    def get_single_schema(self) -> Optional[Schema]:
        return self._single_schema

    def get_tuple_schema(self) -> Optional[List[Schema]]:
        return self._tuple_schema

    def is_single_type(self) -> bool:
        return self._single_schema is not None

    @staticmethod
    def of(*schemas: Schema) -> ArraySchemaType:
        if len(schemas) == 1:
            return ArraySchemaType().set_single_schema(schemas[0])
        return ArraySchemaType().set_tuple_schema(list(schemas))

    @staticmethod
    def from_value(obj) -> Optional[ArraySchemaType]:
        if obj is None:
            return None

        from kirun_py.json.schema.schema import Schema

        if isinstance(obj, list):
            return ArraySchemaType().set_tuple_schema(Schema.from_list_of_schemas(obj) or [])

        if isinstance(obj, dict):
            keys = list(obj.keys())
            if 'singleSchema' in keys:
                s = Schema.from_value(obj['singleSchema'])
                if s:
                    return ArraySchemaType().set_single_schema(s)
            elif 'tupleSchema' in keys:
                return ArraySchemaType().set_tuple_schema(
                    Schema.from_list_of_schemas(obj['tupleSchema']) or []
                )

            x = Schema.from_value(obj)
            if x is None:
                return None
            return ArraySchemaType().set_single_schema(x)

        return None
